Streamline `Compiler`

nnethercote opened the pull request:
A few commits to clean up `Compiler`.

r? @Zoxc

bors commented:
:umbrella: The latest upstream changes (presumably #63827) made this pull request unmergeable. Please resolve the merge conflicts.

Zoxc commented:
Some of the changes here seems questionable (the shuffling of code and the removal of `compile`). This also conflicts with https://github.com/rust-lang/rust/pull/59904.

nnethercote commented:
This is where I express some frustration.

- Your feedback is not actionable. I have spent time trying to improve the quality of the code. Please be more specific in your criticisms, so that I (a) actually know what they are, and (b) can respond to them and/or act on them.
- You have said this conflicts with #59904. That is one of a chain of 5 PRs. None of them have any explanation of what they are doing beyond the PR title. They all date from March, two of them (#59904 and #59338) are closed, and the other two are tagged with `S-blocked` or `S-waiting-on-team`. It is difficult for me to understand the intention and status of these PRs in their current state.
- My motivation for looking at this code was that I want to try moving metadata generation earlier, to see if it improves the effectiveness of pipelining. This requires understanding the main passes of the compiler, which turns out to be far more complex than I expected due to the implicit ordering that occurs in this query-based design. My first attempt had no effect because I didn't realize that passes could be called multiple times, with the latter calls effectively being no-ops. I spent an afternoon diagramming the passes and their dependencies in order to understand how it works. (The diagram is below.) The changes in this PR minimize and simplify dependencies, merge two unnecessarily-separated passes, and remove an inconsistent function. I think they are worthwhile, and I would appreciate a more respectful engagement with them.

![IMG-4423](https://user-images.githubusercontent.com/1940286/64051751-9a558a00-cbbf-11e9-8bcd-f28abca7ede0.JPG)

nnethercote commented:
Here is an updated picture showing things after my changes. The graph is neater partly because I did a better job drawing it, but also partly because it's simpler: it has three fewer boxes (`passes::register_plugins`, `compile`, and `ongoing_codegen` & `link` have been combined into `codegen_and_link`) and fewer arrows.

![IMG-4424](https://user-images.githubusercontent.com/1940286/64056932-620f7500-cbda-11e9-9674-8be808c1dd16.JPG)

nnethercote commented:
#63827 landed over the weekend, which removed one of the `Compiler::compile()` calls. As a result, the last patch is now more compelling.

nnethercote commented:
The `S-waiting-on-author` tag is incorrect, but I don't know how to remove it.

lzutao commented:
@rustbot modify labels: -S-waiting-on-author S-waiting-on-review

nnethercote commented:
r? @oli-obk 

Because they reviewed #56732.

oli-obk commented on src/librustc_driver/lib.rs:
```diff
@@ -383,17 +383,12 @@ pub fn run_compiler(
-
-        // Drop GlobalCtxt after starting codegen to free memory
-        mem::drop(compiler.global_ctxt()?.take());
+        compiler.codegen_and_link()?;
```
having reviewed @Zoxc's PRs it seems to me like only this commit is actually problematic for the querification PRs (like those PRs would have to revert this commit to be doable). As far as I can tell (both from the code and your graphs) it's not that much additional complexity to keep around either.

What do you think?

nnethercote replied:
Thank you for the feedback. Are you referring to #59404? If so:
- If there's a good reason to keep the `ongoing_codegen`/`link` split then I am fine to remove this commit.
- Is there a good reason? Does #59404 and its predecessors use the existing passes because they are there, rather than because they are necessary?

Zoxc replied:
> * If there's a good reason to keep the `ongoing_codegen`/`link` split then I am fine to remove this commit.

The main reason to keep this separate because `ongoing_codegen` needs the `GlobalCtxt` to be around, while we want to free `GlobalCtxt` before linking. When we remove queries from `rustc_interface` this mean this cannot be a combined operation.

Another reason is that it's reasonable for front-ends to do only code generation, but no linking, although that is not yet possible.

nnethercote replied:
I don't understand the point about `GlobalCtxt`. My patch preserves the "free `GlobalCtxt` before linking" behaviour.

As for the second point, I would argue that YAGNI applies.

oli-obk replied:
The problem isn't that your PR doesn't keep the early drop, it's the way the query PR works, where the `GlobalCtxt` lives as long as a closure and is automatically dropped at the end of it. So you can't call the combined function inside the closure and you can't call it outside the closure, thus needing to split it up again.

Zoxc commented on src/librustc_interface/passes.rs:
```diff
@@ -261,9 +260,6 @@ pub fn register_plugins<'a>(
 
-    // If necessary, compute the dependency graph (in the background).
-    compiler.dep_graph_future().ok();
-
```
Moving the dep graph loading to a later point is a performance regression. This is currently done as early as possible, although I'd like for it be done immediately when creating the compiler context, but that requires some changes as currently we need to do this after parsing to pick up the crate name.

nnethercote replied:
The performance run shows that these five commits provide a slightly performance improvement.

Zoxc commented on src/librustc_interface/queries.rs:
```diff
@@ -246,46 +253,19 @@ impl Compiler {
         })
     }
-
-    pub fn compile(&self) -> Result<()> {
```
This is meant as an simple utility function and an example for users of `rustc_interface` of something that mirrors a regular compilation session without all the complexities of the full rustc driver (in `rustc_driver`).

nnethercote replied:
If an example is needed, why not provide it in documentation? Don't leave unnecessary code in the product.

pnkfelix replied:
Do we have CI testable example code snippets in the rustc-guide?

 * Or maybe we could put such a snippet into the rustdoc comments that we process for `rustc`'s generated documentation?

If either of the above is true, then I could go along with @nnethercote 's reasoning here, though then I would want that migration to happen either before this PR, or as part of this PR.

(Otherwise, I'm with @Zoxc; Its hard enough for people to get started using the compilier as a library or hacking on `rustc` itself; simple examples of how to use things like `rustc_interface` are in their own way part of our product...)

nnethercote commented:
@bors try

bors commented:
:hourglass: Trying commit 92a0cbd3c30ce26498739c8f18cba6ab0cd4e7f8 with merge 468729dc29bac7687ebb584b2a711f3f58e41cf8...
<!-- homu: {"type":"TryBuildStarted","head_sha":"92a0cbd3c30ce26498739c8f18cba6ab0cd4e7f8","merge_sha":"468729dc29bac7687ebb584b2a711f3f58e41cf8"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=7613)
Build commit: 468729dc29bac7687ebb584b2a711f3f58e41cf8
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=7613"},"merge_sha":"468729dc29bac7687ebb584b2a711f3f58e41cf8"} -->

nnethercote commented:
@rust-timer build 468729dc29bac7687ebb584b2a711f3f58e41cf8

rust-timer commented:
Success: Queued 468729dc29bac7687ebb584b2a711f3f58e41cf8 with parent ef54f57c5b9d894a38179d09b00610c1b337b086, [comparison URL](https://perf.rust-lang.org/compare.html?start=ef54f57c5b9d894a38179d09b00610c1b337b086&end=468729dc29bac7687ebb584b2a711f3f58e41cf8).

rust-timer commented:
Finished benchmarking try commit 468729dc29bac7687ebb584b2a711f3f58e41cf8, [comparison URL](https://perf.rust-lang.org/compare.html?start=ef54f57c5b9d894a38179d09b00610c1b337b086&end=468729dc29bac7687ebb584b2a711f3f58e41cf8).

nnethercote commented:
I removed the commit that merges `ongoing_codegen` with `link`. I think the other four commits are still worth landing.

nnethercote commented:
I have changed the fourth commit to retain `Compiler::compile`, and added a comment explaining that it's a useful example.

The four commits as they now stand.
- **Don't call `self.parse()` in `Compiler::crate_name()` unless necessary.** Trivial, should be uncontroversial.
- **Remove `lower_to_hir()` call from `prepare_output()`.** Very simple, should be uncontroversial.
- **Move call site of `dep_graph_future()`.** A clear simplification. The perf results above should allay concerns about perf regressions.
- **Add a comment to `Compiler::compile()`.** Trivial, should be uncontroversial.

@oli-obk: can you re-review? Thanks.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 25211894386a34db1639fbd69680e8f7b35ee1a4 has been approved by `oli-obk`

<!-- @bors r=oli-obk 25211894386a34db1639fbd69680e8f7b35ee1a4 -->
<!-- homu: {"type":"Approved","sha":"25211894386a34db1639fbd69680e8f7b35ee1a4","approver":"oli-obk"} -->
